third_party/pcsc-lite/naclport/server_clients_management/src/permissions_checking/user-prompting-checker-unittests.js

```diff
@@ -0,0 +1,288 @@
+/** @const */
+var UserPromptingChecker = PermissionsChecking.UserPromptingChecker;
+/** @const */
+var ignoreArgument = goog.testing.mockmatchers.ignoreArgument;
+
+/** @const */
+var FAKE_APP_1_ID = 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa';
+/** @const */
+var FAKE_APP_1_NAME = 'App Name 1';
+/** @const */
+var FAKE_APP_1_KNOWN_APP = new KnownApp(FAKE_APP_1_ID, FAKE_APP_1_NAME);
+/** @const */
+var FAKE_APP_2_ID = 'bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb';
+/** @const */
+var STORAGE_KEY = 'pcsc_lite_clients_user_selections';
+
+/**
+ * @enum {number}
+ */
+var MockedDialogBehavior = {
+  NOT_RUN: 0,
+  USER_APPROVES: 1,
+  USER_DENIES: 2,
+  USER_CANCELS: 3,
+};
+
+/**
+ * Sets up a mock instance of KnownAppsRegistry.
+ *
+ * This mock instance is only aware of |FAKE_APP_1_ID|.
+ * @param {!goog.testing.MockControl} mockControl
+ */
+function setUpKnownAppsRegistryMock(mockControl) {
+  var mockInstance = {
+    getById: function(id) {
+      if (id == FAKE_APP_1_ID)
+        return goog.Promise.resolve(FAKE_APP_1_KNOWN_APP);
+      return goog.Promise.reject();
+    }
+  };
+  /** @type {?} */
+  var mockedConstructor = mockControl.createConstructorMock(
+      PermissionsChecking, 'KnownAppsRegistry');
+  mockedConstructor().$returns(mockInstance);
+}
+
+/**
+ * Sets up mocks for get() and set() methods of the chrome.storage.local
+ * Extensions API.
+ * @param {!goog.testing.MockControl} mockControl
+ * @param {!goog.testing.PropertyReplacer} propertyReplacer
+ * @param {!Object} fakeInitialData Fake data to be returned from get().
+ * @param {Object} expectedDataToBeWritten If non-null, then the expected data
+ * to be passed into set(); if null, then set() isn't expected to be called.
+ */
+function setUpChromeStorageMock(
+    mockControl, propertyReplacer, fakeInitialData, expectedDataToBeWritten) {
+  propertyReplacer.set(
+      chrome, 'storage',
+      {local: {
+         get: mockControl.createFunctionMock('chrome.storage.local.get'),
+         set: mockControl.createFunctionMock('chrome.storage.local.set')}});
+
+  /** @type {?} */ chrome.storage.local.get;
+  chrome.storage.local.get(STORAGE_KEY, ignoreArgument).$does(
+      function(key, callback) {
+        callback(fakeInitialData);
+      });
+
+  /** @type {?} */ chrome.storage.local.set;
+  if (!goog.isNull(expectedDataToBeWritten))
+    chrome.storage.local.set(expectedDataToBeWritten);
+}
+
+/**
+ * Sets up mocks for the GSC.PopupWindow.Server.runModalDialog method.
+ * @param {!goog.testing.MockControl} mockControl
+ * @param {!MockedDialogBehavior} mockedBehavior If |NOT_RUN|, then the method
+ * is not expected to be called; otherwise, the called method will return the
+ * corresponding result.
+ */
+function setUpDialogMock(mockControl, mockedBehavior) {
+  /** @type {?} */
+  var mockedFunction = mockControl.createMethodMock(
+      GSC.PopupWindow.Server, 'runModalDialog');
+  var mockAction;
+  switch (mockedBehavior) {
+    case MockedDialogBehavior.NOT_RUN:
+      return;
```

Comment:
Should this fail somehow if the dialog is invoked when not expected?

Reply:
It will - by default, a function mock expects to be not called at all.